Statically initialized objects

eeckstein opened the pull request:
Outline heap objects to global variables.
 
 Mainly this is done for array literals.
 This new optimization creates a statically initialized global variable which is the allocated object.
 The alloc_ref instruction is replaced by a global_value instruction.
    
 This optimization gives significant performance improvements for large array literals (e.g. for the ArrayLiteral benchmark).

eeckstein committed:
SIL: allow some literal and aggregate instructions to be constructed without a function context

just refactoring, NFC

eeckstein committed:
SIL: Let the OpenedArchetypesTracker be constructed with a null SILFunction.

NFC

eeckstein committed:
SILPrinter: allow also basic blocks to be a context.

Plus a small refactoring to simplify the code: remove the inner SILPrintFunctionContext struct

eeckstein committed:
SILParser: small refactoring.

Don't pass the basic block to parseSILInstruction.
NFC

eeckstein committed:
SIL: A new representation of static initializers for global variables.

Static initializers are now represented by a list of literal and aggregate instructions in a SILGlobalVariable.
For details see SIL.rst.

This representation is cleaner than what we did so far (point to the initializer function and do some pattern matching).

One implication of that change is that now (a subset of) instructions not necessarily have a parent function.
Regarding the generated code it's a NFC.
Also the swift module format didn't change because so far we don't serializer global variables.

eeckstein committed:
runtime: add a runtime function to initialize the header of a statically allocated object

eeckstein committed:
mangling: Add mangling for outlined global variables

eeckstein committed:
unittests: fix compiler warnings

eeckstein committed:
IRGen: extend ClassLayoutBuilder to create layouts for classes with tail allocated elements

eeckstein committed:
SIL, IRGen: add instructions "object" and "global_value” to support statically initialized objects.

This commit contains:
-) adding the new instructions + infrastructure, like parsing, printing, etc.
-) support in IRGen to generate global object-variables (i.e. "heap" objects) which are statically initialized in the data section.
-) IRGen for global_value which lazily initializes the object header and returns a reference to the object.

For details see the documentation of the new instructions in SIL.rst.

eeckstein commented:
@swift-ci Please test

jckarter commented on docs/SIL.rst:
```diff
@@ -1210,17 +1210,36 @@ Global Variables
+Global variable access is performed by the ``alloc_global``, ``global_addr``
+and ``global_value`` instructions.
 
+A global can have a static initializer if it's initial value can be
```
`it's` -> `its`

jckarter commented on docs/SIL.rst:
```diff
@@ -1210,17 +1210,36 @@ Global Variables
+the global variable is a statically initialized object. In this case the
+variable cannot be used as l-value, i.e. the reference to the object cannot be
+modified. As a consequence the variable cannot be accessed with ``global_addr``
+but only with ``global_value``.
```
Can we model this more generally, by having a `sil_global` track whether it's immutable or not, and only allowing `object` literals and `global_value` accesses on immutable `sil_global`s?

eeckstein replied:
I'll leave this open for a follow-up work. I think this needs some more discussion.

jckarter replied:
Fair enough. Another thing that crossed my mind—is the constraint on `object` strictly necessary? It seems like something we could support generally by lowering to alloc_ref + initialization of the class fields during IRGen, and it might make some analyses easier to be able to represent object instantiations that way.

eeckstein replied:
Interesting idea

jckarter reviewed (approved):
Looks great overall!

eeckstein committed:
GlobalOpt: Outline heap objects to global variables.

Mainly this is done for array literals.
This new optimization creates a statically initialized global variable which is the allocated object.
The alloc_ref instruction is replaced by a global_value instruction.

This optimization can give significant performance improvements for large array literals.

eeckstein committed:
fix typo in SIL.rst

eeckstein commented:
@swift-ci test and merge

eeckstein commented:
@swift-ci smoke test and merge

eeckstein commented:
@swift-ci smoke test and merge

eeckstein commented:
@swift-ci smoke test and merge

eeckstein commented:
@swift-ci smoke test and merge

shahmishal commented:
https://ci.swift.org/job/oss-swift-package-osx/89/

```
/Users/buildnode/jenkins/workspace/oss-swift-package-osx/swift/test/SILOptimizer/static_arrays.swift:55:15: error: expected string not found in input
// CHECK-DAG: integer_literal $Builtin.Int64, 1
              ^
<stdin>:116:2: note: scanning from here
 %0 = integer_literal $Builtin.Int32, 2 // user: %1
 ^
<stdin>:118:7: note: possible intended match here
 %2 = integer_literal $Builtin.Int32, 1 // user: %3
      ^

--

********************
Testing: 0 .. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 
Testing Time: 986.21s
********************
Failing Tests (1):
    Swift(watchsimulator-i386) :: SILOptimizer/static_arrays.swift
```

jpsim commented on docs/SIL.rst:
```diff
@@ -2851,7 +2870,20 @@ global_addr
+
+Returns the value of a global variable which has been previously initialized
+by ``alloc_global``. It is undefined behavior to perform this operation on a
+global variable which has not been initialized, except the global variable
```
Missing "if": "except if the global variable"
